internals-js/src/operations.ts

```diff
@@ -360,30 +360,37 @@ export class Operation {
   }
 
   optimize(fragments?: NamedFragments, minUsagesToOptimize: number = 2): Operation {
+    assert(minUsagesToOptimize >= 1, `Expected 'minUsagesToOptimize' to be at least 1, but got ${minUsagesToOptimize}`)
     if (!fragments) {
       return this;
     }
+
     let optimizedSelection = this.selectionSet.optimize(fragments);
     if (optimizedSelection === this.selectionSet) {
       return this;
     }
 
-    // Optimizing fragments away, and then de-optimizing if it's used less than we want, feels a bit wasteful,
-    // but it's simple and probably don't matter too much in practice (we only call this optimization on the
-    // final compted query plan, so not a very hot path; plus in most case we won't even reach that point
-    // either because there is no fragment, or none will have been optimized away and we'll exit above). We
-    // can optimize later if this show up in profiling though.
-    if (minUsagesToOptimize > 1) {
-      const usages = new Map<string, number>();
-      optimizedSelection.collectUsedFragmentNames(usages);
-      for (const fragment of fragments.names()) {
-        if (!usages.has(fragment)) {
-          usages.set(fragment, 0);
-        }
+    const usages = new Map<string, number>();
+    optimizedSelection.collectUsedFragmentNames(usages);
+    for (const fragment of fragments.names()) {
+      if (!usages.has(fragment)) {
+        usages.set(fragment, 0);
       }
-      const toDeoptimize = mapEntries(usages).filter(([_, count]) => count < minUsagesToOptimize).map(([name]) => name);
-      optimizedSelection = optimizedSelection.expandFragments(toDeoptimize);
     }
+
+    // We re-expand any fragments that is used less than our minimum. Optimizing all fragments to potentially
+    // re-expand some is not entirely optimal, but it's simple and probably don't matter too much in practice
+    // (we only call this optimization on the final computed query plan, so not a very hot path; plus in most
+    // cases we won't even reach that point either because there is no fragment, or none will have been
+    // optimized away so we'll exit above). We can optimize later if this show up in profiling though.
+    //
+    // Also note `toDeoptimize` will always contains the unused fragments, which will allow `expandFragments`
```

Comment:
Note to reviewers: this comment is the reason of the change in this method. Essentially, it removes the prior `minUsagesToOptimize > ` condition so `expandFragments` can "clean up" unused fragments in all cases.